pkg/mover/chart.go

```diff
@@ -236,17 +285,116 @@ func namespacedPath(fullpath, chartName string) string {
+	if cm.targetOfflineTar != "" {
+		return cm.saveOfflineBundle()
+	}
+	return cm.moveChart()
+}
+
+func (cm *ChartMover) saveOfflineBundle() error {
 	log := cm.logger
 
+	tarPath, err := os.MkdirTemp("", "offline-tarball-*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory to build tar: %w", err)
+	}
+	log.Printf("Writing chart at %s/...\n", cm.chart.Metadata.Name)
+	if err := writeChart(cm.chart, filepath.Join(tarPath, cm.chart.Metadata.Name)); err != nil {
+		return fmt.Errorf("failed archiving chart %s: %w", cm.chart.Name(), err)
+	}
+	if err := packImages(tarPath, cm.imageChanges, cm.logger); err != nil {
+		return fmt.Errorf("failed archiving images: %w", err)
+	}
+	log.Printf("Writing hints file %s...\n", HintsFilename)
+	if err := os.WriteFile(filepath.Join(tarPath, HintsFilename), cm.rawHints, defaultTarPermissions); err != nil {
+		return fmt.Errorf("failed to write hints file: %w", err)
+	}
+	log.Printf("Packing all as tarball %s...\n", cm.targetOfflineTar)
+	if err := tarDirectory(tarPath, cm.targetOfflineTar); err != nil {
+		return fmt.Errorf("failed to tar bundle as %s: %w", cm.targetOfflineTar, err)
+	}
+	return os.RemoveAll(tarPath)
+}
+
+func writeChart(chart *chart.Chart, targetDir string) error {
+	if err := os.MkdirAll(targetDir, defaultTarPermissions); err != nil {
+		return fmt.Errorf("failed to create target chart %s: %w", targetDir, err)
+	}
+	for _, file := range chart.Raw {
+		dir := filepath.Dir(file.Name)
+		dirPath := filepath.Join(targetDir, dir)
+		if err := os.MkdirAll(dirPath, defaultTarPermissions); err != nil {
+			return fmt.Errorf("failed to create path %s: %w", dirPath, err)
+		}
+		if err := os.WriteFile(filepath.Join(targetDir, file.Name), file.Data, defaultTarPermissions); err != nil {
+			return fmt.Errorf("failed to write chart's inner file %s: %v", file.Name, err)
+		}
+	}
+	return nil
+}
+
+func packImages(archivePath string, imageChanges []*internal.ImageChange, logger Logger) error {
+	imagesTarball := filepath.Join(archivePath, "images.tar")
+	tagToImage := map[name.Tag]v1.Image{}
+	for _, change := range imageChanges {
+		app := change.ImageReference.Context().Name()
+		version := change.ImageReference.Identifier()
+		fullImageName := fmt.Sprintf("%s:%s", app, version)
+		tag, err := name.NewTag(fullImageName)
+		if err != nil {
+			return fmt.Errorf("failed to create tag %q: %v", fullImageName, err)
+		}
+		if previousImage, ok := tagToImage[tag]; ok {
+			if err := deduplicateByDigest(fullImageName, change.Image, previousImage); err != nil {
+				return err
+			}
+			continue
+		}
+		tagToImage[tag] = change.Image
+		logger.Printf("Processing image %s\n", fullImageName)
+	}
+	logger.Printf("Packing all %d images within images.tar...\n", len(tagToImage))
+	if err := tarball.MultiWriteToFile(imagesTarball, tagToImage); err != nil {
+		return err
+	}
+	return nil
+}
+
+func deduplicateByDigest(name string, current, previous v1.Image) error {
```

Comment:
Can you explain what this does and why it's needed?

Reply:
Doc string added